pkg/cmd/ssh/command.go

```diff
@@ -0,0 +1,460 @@
+
+	"github.com/gardener/gardener/pkg/apis/core/v1alpha1"
+	corev1alpha1helper "github.com/gardener/gardener/pkg/apis/core/v1alpha1/helper"
+	gardencorev1beta1 "github.com/gardener/gardener/pkg/apis/core/v1beta1"
+	v1beta1constants "github.com/gardener/gardener/pkg/apis/core/v1beta1/constants"
+	operationsv1alpha1 "github.com/gardener/gardener/pkg/apis/operations/v1alpha1"
+	"github.com/gardener/gardener/pkg/utils"
+	"github.com/spf13/cobra"
+	corev1 "k8s.io/api/core/v1"
+	networkingv1 "k8s.io/api/networking/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	// SSHBastionUsername is the system username on the bastion host.
+	SSHBastionUsername = "gardener"
+	// SSHNodeUsername is the system username on any of the shoot cluster nodes.
+	SSHNodeUsername = "gardener"
+	// SSHPort is the TCP port on a bastion instance that allows incoming SSH.
+	SSHPort = 22
+)
+
+// NewCommand returns a new ssh command.
+func NewCommand(f util.Factory, o *Options) *cobra.Command {
+	cmd := &cobra.Command{
+		Use:   "ssh",
+		Short: "Establish an SSH connection to a Shoot cluster's node",
+		RunE: func(cmd *cobra.Command, args []string) error {
+			if err := o.Complete(f, cmd, args, o.IOStreams.Out); err != nil {
+				return fmt.Errorf("failed to complete command options: %w", err)
+			}
+			if err := o.Validate(); err != nil {
+				return err
+			}
+
+			return runCommand(f, o)
+		},
+	}
+
+	cmd.Flags().BoolVar(&o.Interactive, "interactive", o.Interactive, "Open an SSH connection instead of just providing the bastion host.")
+	cmd.Flags().StringArrayVar(&o.CIDRs, "cidr", nil, "CIDRs to allow access to the bastion host; if not given, the host's public IP is auto-detected.")
+	cmd.Flags().StringVar(&o.SSHPublicKeyFile, "public-key-file", "", "Path to the file that contains a public SSH key. If not given, a temporary keypair will be generated.")
+	cmd.Flags().DurationVar(&o.WaitTimeout, "wait-timeout", o.WaitTimeout, "Maximum duration to wait for the bastion to become available.")
+
+	return cmd
+}
+
+func runCommand(f util.Factory, o *Options) error {
+	manager, err := f.Manager()
+	if err != nil {
+		return err
+	}
+
+	// validate the current target
+	currentTarget, err := manager.CurrentTarget()
+	if err != nil {
+		return err
+	}
+
+	if currentTarget.ShootName() == "" {
+		return errors.New("no Shoot cluster targeted")
+	}
+
+	// create client for the garden cluster
+	gardenClient, err := manager.GardenClient(currentTarget)
+	if err != nil {
+		return err
+	}
+
+	// fetch targeted shoot (ctx is cancellable to stop the keep alive goroutine later)
+	ctx, cancel := context.WithCancel(context.Background())
+	defer cancel()
+
+	shoot, err := util.ShootForTarget(ctx, gardenClient, currentTarget)
+	if err != nil {
+		return err
+	}
+
+	// fetch the SSH key for the shoot nodes
+	nodePrivateKey, err := getShootNodePrivateKey(ctx, gardenClient, shoot)
```

Comment:
also fetch the old key in case there is one and handle fallback in case the ssh connection fails with the new key (e.g. `authorized_keys` file on the node was not yet updated).

Reply:
Do we really want to implement something here that doesn't exist yet in Gardener?

Reply:
then let's do it in a follow up pr